modules/lotameIdSystem.js

```diff
@@ -0,0 +1,178 @@
+    }
+    if (storage.hasLocalStorage()) {
+      storage.setDataInLocalStorage(key, value, undefined);
+    }
+  }
+}
+
+function getLotameLocalCache() {
+  let cache = {
+    lastUpdate: new Date(0),
+    data: getFromStorage(KEY_ID),
+    bundle: {
+      refreshSeconds: 3600,
+    },
+  };
+
+  try {
+    const rawBundle = getFromStorage(KEY_EXPIRY);
+    if (utils.isStr(rawBundle)) {
+      cache.bundle = JSON.parse(rawBundle);
+    }
+  } catch (error) {
+    utils.logError(error);
+  }
+  try {
+    const rawDate = getFromStorage(KEY_LAST_UPDATE);
+    if (utils.isStr(rawDate)) {
+      cache.lastUpdate = new Date(rawDate);
+    }
+  } catch (error) {
+    utils.logError(error);
+  }
+  return cache;
+}
+
+function clearLotameCache(key) {
+  if (key) {
+    if (storage.cookiesAreEnabled()) {
+      let expirationDate = new Date(0).toUTCString();
+      storage.setCookie(key, '', expirationDate, 'Lax', undefined, undefined);
+    }
+    if (storage.hasLocalStorage()) {
+      storage.removeDataFromLocalStorage(key, undefined);
+    }
+  }
+}
+
+/** @type {Submodule} */
+export const lotameIdSubmodule = {
+  /**
+   * used to link submodule with config
+   * @type {string}
+   */
+  name: MODULE_NAME,
+
+  /**
+   * Decode the stored id value for passing to bid requests
+   * @function decode
+   * @param {(Object|string)} value
+   * @returns {(Object|undefined)}
+   */
+  decode(value, configParams) {
+    return value;
+  },
+
+  /**
+   * Retrieve the Lotame id
+   * @function
+   * @param {SubmoduleParams} [configParams]
+   * @param {ConsentData} [consentData]
+   * @param {(Object|undefined)} cacheIdObj
+   * @returns {IdResponse|undefined}
+   */
+  getId(configParams, consentData, cacheIdObj) {
+    let localCache = getLotameLocalCache();
+
+    let refreshNeeded = false;
+    if (!utils.isStr(localCache.data)) {
+      const lastUpdate = localCache.lastUpdate;
+      refreshNeeded = (Date.now() - lastUpdate.getTime()) > localCache.bundle.refreshSeconds * 1000;
+    }
+
+    if (!refreshNeeded) {
+      return {};
+    }
+
+    const storedUserId = getFirstPartyId();
+    const resolveIdFunction = function (callback) {
+      const param = storedUserId ? `?profile_id=${storedUserId}` : '';
+      const url = 'https://mconrad.dev.lotame.com:5555/panorama/id' + param;
+      ajax(
+        url,
+        (response) => {
+          let responseObj = {};
+          if (response) {
+            try {
+              responseObj = JSON.parse(response);
+              setFirstPartyId(responseObj.profile_id);
+              saveLotameCache(KEY_LAST_UPDATE, new Date().toUTCString());
+              if (utils.isStr(responseObj.panorama_id)) {
+                saveLotameCache(KEY_ID, responseObj.panorama_id);
+              } else {
+                clearLotameCache(KEY_ID);
+              }
+              // TODO: Get this from the response
+              saveLotameCache(KEY_EXPIRY, JSON.stringify({
```

Comment:
What is this?

Reply:
The information that we will eventually get back from the endpoint on how long to wait for a refresh